File: problem5.py
from typing import Iterable, Tuple, List, NewType, TypeVar, Callable, Any
from functools import reduce
import operator

# prereq functions
# myFoldl function
def myFoldl(f: Callable[[Any, Any], Any], accInit: Any, values: List[Any]) -> Any:
    """foldl using 10 as our initializer and add the numbers in the list
    >>> myFoldl(operator.add, 10, [1, 2, 3, 4, 5])
    25
    """
    return reduce(lambda x, y: f(f(accInit, x), y) if x == values[0] else f(x,y), values)

#
def reverse(values: List[int]) -> List[int]:
    return values[::-1]

# define a sequence
Section = NewType('Section', Tuple[int, int, int])
Path = NewType('Path', List[Tuple[str, int]])
RoadSystem = NewType('RoadSystem', List[Section])

# define heathrow to london as a group of sections
heathrowToLondon = RoadSystem([Section((50, 10, 30)), Section((5, 90, 20)), Section((40, 2, 25)), Section((10, 8, 0))])

# define our roadStepFast function
def roadStepFast(pathData: Tuple[Path, Path, int, int], sectionData: Section, debug: bool = True) -> Tuple[Path, Path, int, int]:
    sectionM = {"a" : sectionData[0], "b" : sectionData[1], "c" : sectionData[2]}
    forwardPriceToA = pathData[2] + sectionM['a']
    crossPriceToA = pathData[3] + sectionM['b'] + sectionM['c']
    forwardPriceToB = pathData[3] + sectionM['b']
    crossPriceToB = pathData[2] + sectionM['a'] + sectionM['c']
    if forwardPriceToA <= crossPriceToA:
        newPathToA = Path([("A", sectionM['a']), *pathData[0]])
        newPriceA = forwardPriceToA
    else:
        newPathToA = Path([("C", sectionM['c']), ("B", sectionM['b']), *pathData[0]])
        newPriceA = crossPriceToA
    if forwardPriceToB <= crossPriceToB:
        newPathToB = Path([("B", sectionM['b']), *pathData[1]])
        newPriceB = forwardPriceToB
    else: 
        newPathToB = Path([("C", sectionM['c']), ["A", sectionM['a']], *pathData[0]])
        newPriceB = crossPriceToB
    if debug:
        print(f'PathToA: {newPathToA}\nPathToB: {newPathToB} \nPriceA: {newPriceA} \nPriceB: {newPriceB}')
    return (newPathToA, newPathToB, newPriceA , newPriceB)

# define our optimalPath function
def optimalPath(system: RoadSystem, debug: bool = False) -> Path:
    (bestAPath, bestBPath, priceA, priceB) = myFoldl(roadStepFast, ([], [], 0, 0), system)
    if priceA <= priceB:
        return reverse(bestAPath)
    else:
        return reverse(bestBPath)

if __name__ == "__main__":
    print(optimalPath(heathrowToLondon, True))

File: problem5.hs
-- import trace
import Debug.Trace
    
-- define our data and types
data Section = Section { getA :: Int, getB :: Int, getC :: Int } deriving (Show)  
type RoadSystem = [Section]
data Label = A | B | C deriving (Show)  
type Path = [(Label, Int)]  

-- define the heathrow to london path as a road system
heathrowToLondon :: RoadSystem  
heathrowToLondon = [Section 50 10 30, Section 5 90 20, Section 40 2 25, Section 10 8 0]

-- define our roadStepFast function, this function is has an optimization by having int parameters storing the price of a path
roadStepFast ::(Path, Path, Int, Int) -> Section -> (Path, Path, Int, Int)
roadStepFast (pathA, pathB, priceA, priceB) (Section a b c) =
    let forwardPriceToA =  (priceA + a)
        crossPriceToA =  (priceB + b + c)
        forwardPriceToB =  (priceB + b)
        crossPriceToB =  (priceA + a + c)
        newPathToA = if forwardPriceToA <= crossPriceToA  
            then (A,a):pathA  
            else (C,c):(B,b):pathB  
        newPathToB = if forwardPriceToB <= crossPriceToB  
            then (B,b):pathB  
            else (C,c):(A,a):pathA
        newPriceA = if forwardPriceToA <= crossPriceToA
            then forwardPriceToA
            else crossPriceToA
        newPriceB = if forwardPriceToB <= crossPriceToB
            then forwardPriceToB
            else crossPriceToB
    in (newPathToA, newPathToB, newPriceA, newPriceB)

-- define a wrapper for roadstep fast to perform debugging
roadStepFast' ::(Path, Path, Int, Int) -> Section -> (Path, Path, Int, Int)
roadStepFast' (pathA, pathB, priceA, priceB) (Section a b c) =
    let (pathAResult, pathBResult, priceAResult, priceBResult) = roadStepFast (pathA, pathB, priceA, priceB) (Section a b c)
    in trace ("PathToA: " ++ show pathAResult ++ "\n" ++ "PathToB: " ++ show pathBResult ++ "\n")
        (pathAResult, pathBResult, priceAResult, priceBResult)

-- define our roadStep function, this gets the lowest price path to A or B, not optimized since we calculate the price everytime
roadStep :: (Path, Path) -> Section -> (Path, Path)  
roadStep (pathA, pathB) (Section a b c) =   
    let priceA = sum $ map snd pathA  
        priceB = sum $ map snd pathB 
        forwardPriceToA = priceA + a  
        crossPriceToA = priceB + b + c  
        forwardPriceToB = priceB + b  
        crossPriceToB = priceA + a + c  
        newPathToA = if forwardPriceToA <= crossPriceToA  
                        then (A,a):pathA  
                        else (C,c):(B,b):pathB  
        newPathToB = if forwardPriceToB <= crossPriceToB  
                        then (B,b):pathB  
                        else (C,c):(A,a):pathA  
    in (newPathToA, newPathToB)  

-- define a wrapper for roadstep to perform debugging
roadStep' ::(Path, Path) -> Section -> (Path, Path)
roadStep' (pathA, pathB) (Section a b c) =
    let (pathAResult, pathBResult) = roadStep (pathA, pathB) (Section a b c)
    in trace ("PathToA: " ++ show pathAResult ++ "\n" ++ "PathToB: " ++ show pathBResult ++ "\n")
        (pathAResult, pathBResult)
       
-- define the optimal path function which foldl's the results of roadStep
-- the Bool parameter is to enable or disable debugging
optimalPath :: RoadSystem -> Bool -> Path  
optimalPath roadSystem debug
    | debug =  
        let (bestAPath, bestBPath) = foldl roadStep' ([],[]) roadSystem  
        in  if sum (map snd bestAPath) <= sum (map snd bestBPath)  
                then reverse bestAPath  
                else reverse bestBPath
    | otherwise =
        let (bestAPath, bestBPath) = foldl roadStep ([],[]) roadSystem  
        in  if sum (map snd bestAPath) <= sum (map snd bestBPath)  
                then reverse bestAPath  
                else reverse bestBPath

-- define the optimal path function which foldl's the results of roadStepFast
-- the Bool parameter is to enable or disable debugging
optimalPathFast :: RoadSystem -> Bool -> Path  
optimalPathFast roadSystem debug
    | debug =
        let (bestAPath, bestBPath, priceA, priceB) = foldl roadStepFast' ([],[], 0, 0) roadSystem  
        in  if priceA <= priceB  
                then reverse bestAPath  
                else reverse bestBPath
    | otherwise =
        let (bestAPath, bestBPath, priceA, priceB) = foldl roadStepFast ([],[], 0, 0) roadSystem  
        in  if priceA <= priceB  
                then reverse bestAPath  
                else reverse bestBPath

